### PR TITLE
Design Doc: Vertical Fallback in Data Loading

### LINKED ISSUE #1203: Document design choices for vertical fallback and deduplication in data loading

### PR DESCRIPTION
**Pull request by sffc**
Fixes #1203

I've been promising a design doc for the vertical fallback implementation, a follow-up to Mihai's doc from earlier in the quarter about the theory.

The doc assumes that the reader is already familiar with the subject matter; I do not spend much time clarifying basic concepts.

In the doc, I state some claims and assumptions.  At the end, I lay out four possible implementation approaches:

1. Complete Runtime Vertical Fallback
2. Simple Vertical Fallback with Minimized Locales
3. Simple Vertical Fallback with Manually Maximized Locales
4. Simple Vertical Fallback with Automatically Maximized Locales

I claim that *all of these approaches achieve the same end result*, but with their own caveats.  I'd appreciate feedback on these four approaches, but try to keep feedback on the technical merits.

### CODE REVIEW

**Comment by sffc** [comment]
CC @richgillam

File: docs/design/vertical_fallback.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+**Pros:**
+
+- Scales well to many data keys
+- Shares code and data with an existing component, rather than introducing a novel chunk of code and data as in the Runtime Approach
+
+**Cons:**
+
+- Requires shipping the likely subtags code and data
+- All lookups incur a small but nonempty penalty when maximization occurs
```
Comment (richgillam) [inline docs/design/vertical_fallback.md:163]: I said my piece in the ICU-TC meeting the other day, but I think I vote for Option 3, with a provision to fall back to Option 2 for implementations that know their locale data doesn't need it and don't want to incur the space penalty for the likely-subtags tables.

File: docs/design/vertical_fallback.md [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+
+**Pros:**
+
+- The industry tends to use minimized locales by default (`zh-TW`, `sr-ME`, …), so this option requires less pre-processing on locales at runtime.
+
+**Cons:**
+
+- Footgun: If a non-minimized locale is passed at runtime, it could have the incorrect behavior. For example, `sr-Latn-ME` or `de-Latn-LI` may not find the correct data.
+- Struggles to scale when there are many data keys
+
```
Comment (richgillam) [inline docs/design/vertical_fallback.md:132]: These cons sound disqualifying to me.

**Review by richgillam — COMMENTED** [review]
I commented on this in the ICU-TC meeting yesterday, and I don't really have anything new to say.  I wanted to reread this document to make sure I didn't miss anything, and I think I understand what you're proposing a lot better now that I've read it a second time, and I think I like what you're doing here.

As I mentioned, I'm interested in seeing what you do about _horizontal_ inheritance, but I know that's a separate problem...

**Comment by sffc** [comment]
Thank you @richgillam for your review!

This is a friendly ping to all my other reviewers:

- @dminor as the likely subtags expert
- @zbraniecki as the author of the locale class
- @mihnita as the author of the language negotiation doc
- @nciric for visibility into size-constrained use cases
- @markusicu as the primary implementer of this logic in ICU4C

**Review by dminor — APPROVED** [review]
This all seems reasonable to me. I agree with Rich's comment, option 3, with a way to disable maximization for users that don't want it seems like a good way ahead.

**Comment by sffc** [comment]
I will prepare an update to this PR within the next week and then re-request review.

**Comment by sffc** [comment]
This PR is obsolete so I will close it and replace it with a new one later.

**Pull request closed without merging**